File: knhmp/src/main/java/io/github/fopwoc/knhmp/KnhMpExtension.java
package io.github.fopwoc.knhmp;

import org.gradle.api.Action;
import org.gradle.api.Project;

public class KnhMpExtension {

    private final Project project;
    private final KnhMpTargets targets;
    private final KnhMpSourceSets sourceSets = new KnhMpSourceSets();
    private final KnhMpBuildScope common = new KnhMpBuildScope();
    private String modId;
    private String modName;
    private String modGroup;
    private String modVersion;
    private int javaToolchain = 26;

    public KnhMpExtension(Project project) {
        this.project = project;
        targets = new KnhMpTargets();
        modId = stringProperty("modId", project.getName().replace("-", ""));
        modName = stringProperty("modName", modId);
        modGroup = stringProperty("modGroup", "io.github.example." + modId);
        modVersion = stringProperty("modVersion", "0.1.0");
    }

    public void targets(Action<? super KnhMpTargets> action) {
        action.execute(targets);
    }

    public void sourceSets(Action<? super KnhMpSourceSets> action) {
        action.execute(sourceSets);
    }

    /** Build plugins applied to every target and variant of this module. */
    public void plugins(Action<? super KnhMpPlugins> action) {
        action.execute(common.getPlugins());
    }

    /** Dependencies of every target and variant of this module, e.g. {@code module(":framework")}. */
    public void dependencies(Action<? super KnhMpDependencies> action) {
        action.execute(common.getDependencies());
    }

    public KnhMpBuildScope getCommon() {
        return common;
    }

    public KnhMpTargets getTargets() {
        return targets;
    }

    public KnhMpSourceSets getSourceSets() {
        return sourceSets;
    }

    public String getModId() {
        return modId;
    }

    public void setModId(String modId) {
        this.modId = modId;
    }

    public String getModName() {
        return modName;
    }

    public void setModName(String modName) {
        this.modName = modName;
    }

    public String getModGroup() {
        return modGroup;
    }

    public void setModGroup(String modGroup) {
        this.modGroup = modGroup;
    }

    public String getModVersion() {
        return modVersion;
    }

    public void setModVersion(String modVersion) {
        this.modVersion = modVersion;
    }

    public int getJavaToolchain() {
        return javaToolchain;
    }

    public void setJavaToolchain(int javaToolchain) {
        this.javaToolchain = javaToolchain;
    }

    private String stringProperty(String name, String fallback) {
        Object value = project.findProperty(name);
        return value == null ? fallback : value.toString();
    }
}

File: knhmp/src/main/kotlin/io/github/fopwoc/knhmp/KnhMpExtension.kt
package io.github.fopwoc.knhmp

import org.gradle.api.Action
import org.gradle.api.Project

open class KnhMpExtension(private val project: Project) {

    val targets = KnhMpTargets()
    val sourceSets = KnhMpSourceSets()
    val common = KnhMpBuildScope()
    val repositories = KnhMpRepositories()

    var modId = stringProperty("modId", project.name.replace("-", ""))
    var modName = stringProperty("modName", modId)
    var modGroup = stringProperty("modGroup", "io.github.example.$modId")
    var modVersion: String
        get() = explicitModVersion ?: KnhMpBuildIdentity.version(project.rootDir)
        set(value) {
            explicitModVersion = value
        }
    private var explicitModVersion: String? = project.findProperty("modVersion")?.toString()

    /** Web page of the source repository: the `repositoryUrl` property, else derived from the `origin` remote. */
    var repositoryUrl: String
        get() = explicitRepositoryUrl ?: KnhMpBuildIdentity.repositoryUrl(project.rootDir)
        set(value) {
            explicitRepositoryUrl = value
        }
    private var explicitRepositoryUrl: String? = project.findProperty("repositoryUrl")?.toString()

    /** Jar base name, e.g. `knh-core` produces `knh-core-gtnh-<version>.jar`. */
    var archiveName: String = project.name
    var javaToolchain = 26
    internal val jvmTargetExceptions = mutableMapOf<String, Int>()

    /** A generated class intentionally compiled for an older JVM than the target's main sources. */
    fun jvmTargetException(className: String, target: Int) {
        require(target >= 8) { "JVM target must be at least 8" }
        jvmTargetExceptions[className.replace('.', '/') + ".class"] = target
    }

    fun targets(action: Action<in KnhMpTargets>) = action.execute(targets)

    /** Repositories every compiler island of this module and of its dependents resolves from. */
    fun repositories(action: Action<in KnhMpRepositories>) = action.execute(repositories)

    fun sourceSets(action: Action<in KnhMpSourceSets>) = action.execute(sourceSets)

    /** Build plugins applied to every target and variant of this module. */
    fun plugins(action: Action<in KnhMpPlugins>) = action.execute(common.plugins)

    /** Dependencies of every target and variant of this module, e.g. `module(":framework")`. */
    fun dependencies(action: Action<in KnhMpDependencies>) = action.execute(common.dependencies)

    private fun stringProperty(name: String, fallback: String): String =
        project.findProperty(name)?.toString() ?: fallback
}
